roxctl/common/environment/auth_config.go

```diff
@@ -0,0 +1,143 @@
+)
+
+const (
+	refreshTokenPath = "/sso/session/tokenrefresh"
+)
+
+type configMethod struct {
+	env Environment
+}
+
+var (
+	_ auth.Method = (*configMethod)(nil)
+)
+
+// ConfigMethod provides an auth.Method for using authentication via local configuration.
+// It will use the configuration store by roxctl central login.
+func ConfigMethod(env Environment) auth.Method {
+	return &configMethod{
+		env: env,
+	}
+}
+
+func (c configMethod) Type() string {
+	return "local configuration"
+}
+
+func (c configMethod) GetCredentials(url string) (credentials.PerRPCCredentials, error) {
+	token, err := c.retrieveToken(url)
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving token")
+	}
+	return tokenbased.PerRPCCredentials(token), nil
+}
+
+func (c configMethod) retrieveToken(url string) (string, error) {
+	cfgStore, err := c.env.Config()
+	if err != nil {
+		return "", errors.Wrap(err, "retrieving config store")
+	}
+	cfg, err := cfgStore.Read()
+	if err != nil {
+		return "", errors.Wrap(err, "reading configuration")
+	}
+
+	access := cfg.GetCentralConfigs().GetCentralConfig(url).GetAccess()
+	if access == nil || (access.AccessToken == "" && access.RefreshToken == "") {
+		return "", errox.NoCredentials.Newf(
+			`no credentials found for %s, please run "roxctl central login" to obtain credentials`, url)
+	}
+
+	// 1. Check if an access token is given. If an access token is given, and it's not yet expired, then return it.
+	if access.AccessToken != "" && access.ExpiresAt != nil && access.ExpiresAt.After(time.Now().UTC().Add(30*time.Second)) {
+		return access.AccessToken, nil
+	}
+
+	// 2. In case an access token is given, but either no expiration is indicated or the token is already expired, and
+	// no refresh token is available, print a warning that the authentication _may_ fail and users have to re-run
+	// roxctl central login in case of authentication issues.
+	if access.AccessToken != "" && access.RefreshToken == "" {
+		c.env.Logger().WarnfLn(`No indication about expiration found or already expired for access token for central %s.
+Still, trying to authenticate with the given token. In case there's any issues, run "roxctl central login" again.`, url)
+		return access.AccessToken, nil
+	}
+
+	// 3. This is the case where an access token is either not set or expired, and a refresh token is available.
+	// We attempt to refresh the access token here.
+	if err := c.refreshAccessToken(url, access); err != nil {
+		c.env.Logger().WarnfLn("An error occurred during access token refresh. Try running roxctl central login again.")
+		return "", errors.Wrap(err, "refreshing access token")
+	}
+
+	// 4. Write the config with the updated data to the store.
+	if err := cfgStore.Write(cfg); err != nil {
+		return "", errors.Wrap(err, "writing configuration to store after refreshing the token")
+	}
+
+	return access.AccessToken, nil
+}
+
+func (c configMethod) refreshAccessToken(url string, accessConfig *config.CentralAccessConfig) error {
```

Comment:
Hm, not really sure what you mean by that.

By re-use the information from the created request, which one do you mean? The request created by the HTTP client that will retrieve the token via the auth method, essentially changing the interface for auth method?

I'm not sure it'll be possible, since we have to support both GRPC and HTTP clients, and the GRPC one only relies on the endpoint+port combination.

Reply:
Ah, I _think_ I understand now.

However, the `url` is we use here is actually already the URL from the HTTP client we pass to `GetCredentials` within [here](https://github.com/stackrox/stackrox/pull/5752/files#diff-c571d80a0e339a61310a8c3e85b5217acdc90708e6d8c8b325be55500919e4c7R119), and that's already the hostname + port. So, it's not really the URL extracted from the roxctl instance config, but rather from the HTTP client we have here, unless I'm not understanding the point.

I'll leave it as-is for now, I can address things post merge as well.